### PR TITLE
Analog Continuous mode API

### PR DESCRIPTION
**Pull request by P-R-O-C-H-Y**
## Description of Change
Added new API for running ADC peripheral in continuous mode, which enables ADC conversions to be non-blocking.

Updated documentation with new API + Added new example AnalogReadContinuous to demonstrate the usage.

## Tests scenarios
Tested on all supported chips.

## Related links
Related to  #5205

### CODE REVIEW

File: docs/source/api/adc.rst [hunk 1]
```diff
@@ -16,8 +16,8 @@ ADCs are very useful in control and monitoring applications since most sensors
 Arduino-ESP32 ADC API
 ---------------------
 
-ADC common API
-**************
+ADC OneShot mode
+****************
```
Comment (pedrominatel) [inline docs/source/api/adc.rst:21]: I think we should explain the basic differences between both modes.

**Review by pedrominatel — COMMENTED** [review]
Just one point so far, @P-R-O-C-H-Y.

File: docs/source/api/adc.rst (hunk 1 above)
Comment (P-R-O-C-H-Y) [reply docs/source/api/adc.rst:21]: @pedrominatel Updated. Is this explanation understandable and enough? Thanks

File: libraries/ESP32/examples/AnalogReadContinuous/AnalogReadContinuous.ino [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    analogContinuous(adc_pins, adc_pins_count, CONVERSIONS_PER_PIN, 20000, &adcComplete);
+
+    // Start ADC Continuous conversions
+    analogContinuousStart();
+}
+
+void loop() {
+    // Check if conversion is done and try to read data
+    if (adc_coversion_done == true) {
+    adc_coversion_done = false;
```
Comment (me-no-dev) [inline libraries/ESP32/examples/AnalogReadContinuous/AnalogReadContinuous.ino:47]: you have a styling issue with this whole block. Please fix the spacing

File: libraries/ESP32/examples/AnalogReadContinuous/AnalogReadContinuous.ino [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+    // Read data from ADC
+    if (!analogContinuousRead(&result, 0)) {
```
Comment (me-no-dev) [inline libraries/ESP32/examples/AnalogReadContinuous/AnalogReadContinuous.ino:50]: Should this check if `analogContinuousRead` returned false? Typo?

File: libraries/ESP32/examples/AnalogReadContinuous/AnalogReadContinuous.ino [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    }
+    else {
+        Serial.println("Error occured during reading data. Set Core Debug Level to error or lower for more informations.");
+    }
+
+    // Delay for better readability of ADC data
+    delay(1000);
+
+    // Optional: Start ADC conversions and wait for callback function to set adc_coversion_done to true
+    analogContinuousStart();
```
Comment (me-no-dev) [inline libraries/ESP32/examples/AnalogReadContinuous/AnalogReadContinuous.ino:69]: if reading failed, `analogContinuousStop` was not called, so should you start it here in that case?

File: cores/esp32/esp32-hal-adc.c [hunk 5]
```diff
@@ -310,4 +336,357 @@ extern void analogSetPinAttenuation(uint8_t pin, adc_attenuation_t attenuation)
...
+        esp_err_t err = adc_continuous_read(adc_handle[ADC_UNIT_1].adc_continuous_handle, adc_read, adc_handle[0].conversion_frame_size, &bytes_read, timeout_ms);
+        if(err != ESP_OK){
+            if(err == ESP_ERR_TIMEOUT){
+                log_e("Reading data failed: No data, increase timeout");
+            }
+            else {
+                log_e("Reading data failed with error: %X", err);
+            }
+            *buffer = NULL;
+            return ESP_FAIL;
```
Comment (me-no-dev) [inline cores/esp32/esp32-hal-adc.c:595]: functions marked as returning bool, but you are returning esp_err_t in them?

Comment (P-R-O-C-H-Y) [reply cores/esp32/esp32-hal-adc.c:595]: @me-no-dev will it be better for users to return bool or esp_err_t?

Comment (me-no-dev) [reply cores/esp32/esp32-hal-adc.c:595]: bool is OK for Arduino

**Review by pedrominatel — COMMENTED** [review]
Just some minor improvements.

File: docs/source/api/adc.rst [hunk 6]
```diff
@@ -135,27 +137,140 @@ This function is used to set the attenuation for a specific pin/ADC channel. For
...
 Range is 9 - 12.
 
 .. code-block:: arduino
 
     void analogSetWidth(uint8_t bits);
+
+ADC Continuous mode
+*******************
+
+ADC Continuous mode is a new API for doing analog conversions on multiple pins in background and get a callback when those conversions are done to read the results.
```
[the remark this answers is not in the document]
Comment (P-R-O-C-H-Y) [reply docs/source/api/adc.rst]: Replaced with yours. Thank you Pedro :)

File: docs/source/api/adc.rst [hunk 7]
```diff
@@ -11,13 +11,15 @@ to a digital form so that it can be read and processed by a microcontroller.
...
 
 Arduino-ESP32 ADC API
 ---------------------
 
-ADC common API
-**************
+ADC OneShot mode
+****************
+
+ADC OneShot mode API is fully compatible with the Arduino's ``analogRead``. When you call ``analogRead`` or ``analogReadMillivolts`` funtion will return a result of a single conversion on requested pin.
```
[the remark this answers is not in the document]
Comment (P-R-O-C-H-Y) [reply docs/source/api/adc.rst]: Replaced with yours. Thank you Pedro :)

**Comment by zuo785843091** [comment]
The following files are not found when compiling:
"esp_adc/adc_oneshot.h"
"esp_adc/adc_continuous.h"
"esp_adc/adc_cali_scheme.h"

File: libraries/ESP32/examples/AnalogReadContinuous/AnalogReadContinuous.ino [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+            delay(1000);
+
+            // Optional: If ADC was stopped, start ADC conversions and wait for callback function to set adc_coversion_done flag to true
+            analogContinuousStart();
+        }
+        else {
+            Serial.println("Error occured during reading data. Set Core Debug Level to error or lower for more informations.");
+        }
+    }
+}
```
Comment (lucasssvaz) [inline libraries/ESP32/examples/AnalogReadContinuous/AnalogReadContinuous.ino:71]:
```suggestion
}

```

**Review by lucasssvaz — APPROVED** [review]
Tested on ESP32. Works as expected.
Just a suggestion to add a newline at the end of the .INO so the file is a valid text/source file in POSIX systems.

**Pull request closed without merging**